src/HTTPure/Body.purs

```diff
@@ -1,25 +1,64 @@
 module HTTPure.Body
   ( Body
+  , string
+  , binary
+  , isString
+  , isBinary
+  , fromString
+  , fromBinary
   , read
   , write
+  , size
   ) where
 
 import Prelude
 
 import Data.Either as Either
+import Data.String as String
 import Effect as Effect
 import Effect.Aff as Aff
 import Effect.Ref as Ref
+import Node.Buffer as Buffer
 import Node.Encoding as Encoding
 import Node.HTTP as HTTP
 import Node.Stream as Stream
 
 -- | The `Body` type is just sugar for a `String`, that will be sent or received
 -- | in the HTTP body.
-type Body = String
+data Body
+  = StringBody String
+  | BinaryBody Buffer.Buffer
+
+-- | Construct a `Body` from a string
+string :: String -> Body
+string = StringBody
+
+-- | Construct a `Body` from a `Buffer`
+binary :: Buffer.Buffer -> Body
+binary = BinaryBody
```

Comment:
It's mostly a convention. If data types are opaque, it's easier to change their internal implementation without breaking the public API.

You have good points, though, so in this case it's probably best to just expose the constructors to get rid of the hassle it causes :)

Reply:
I see what you're getting at; totally makes sense.  But yeah, in this case, I still think it's better to expose the constructors, mostly since doing things with constructors enables pattern matching which would be a super useful construct for working with different body types.